crates/turborepo-lib/src/rewrite_json.rs

```diff
@@ -0,0 +1,411 @@
+
+#[derive(Error, Debug)]
+pub enum RewriteError {
+    #[error("The JSON document failed to parse. {0}")]
+    ParseError(ParseError),
+    #[error("The JSON document contains no root object.")]
+    NoRoot,
+}
+/**
+ * When generating replacement content it's one of two things.
+ */
+enum GenerateType {
+    Object,
+    Member,
+}
+
+/**
+ * Range stores the primitive AST token details required for document
+ * mutation.
+ */
+struct Range {
+    pub start: usize,
+    pub end: usize,
+    pub replacement_char: String,
+}
+
+/**
+ * Given a JSONC document, an object traversal path, and a _pre-serialized_
+ * JSON value `set_path` will return a minimally-mutated JSONC document with
+ * the path specified set to the JSON value.
+ *
+ * - If the path exists, it will clobber the existing contents.
+ * - If the path does not exist, it will synthesize object members and
+ *   objects to ensure it does exist.
+ *
+ * In the event that the key appears multiple times in the document the last
+ * instance of the key will be updated and earlier instances will be
+ * disregarded.
+ */
+pub fn set_path(
+    json_document_string: &str,
+    path: Vec<&str>,
+    json_value: &str,
+) -> Result<String, RewriteError> {
+    let root = get_root(json_document_string)?;
+
+    // Find the token we'll be modifying and its path from the root.
+    let (closest_path, closest_node) = get_closest_node(&root, &path, vec![]);
+
+    // Pull the token metadata off of the token.
+    let (property_count, range): (usize, jsonc_parser::common::Range) = match closest_node {
+        jsonc_parser::ast::Value::Object(literal) => (literal.properties.len(), literal.range),
+        jsonc_parser::ast::Value::StringLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NumberLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::BooleanLit(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::Array(literal) => (0, literal.range),
+        jsonc_parser::ast::Value::NullKeyword(literal) => (0, literal.range),
+    };
+
+    // Figure out what we should be generating:
+    // - An object to be assigned to an existing member. ("object")
+    // - A member to add to an existing object. ("member")
+    let generate_type: GenerateType = if !closest_path.is_empty() {
+        GenerateType::Object
+    } else {
+        GenerateType::Member
+    };
+
+    // Identify the token replacement metadata: start, end, and possible trailing
+    // join character
+    let separator;
+    let start;
+    let end;
+
+    match generate_type {
+        GenerateType::Object => {
+            separator = "";
+            start = range.start;
+            end = range.end;
+        }
+        GenerateType::Member => {
+            if property_count > 0 {
+                separator = ",";
+            } else {
+                separator = ""
+            }
+
+            start = range.start + 1;
+            end = range.start + 1;
+        }
+    }
```

Comment:
I think it's maybe a bit less-vulnerable to mistakes to do it this way, but less-clear:

```rust
    let (separator, start, end) = match generate_type {
        GenerateType::Object => {
            ("", range.start, range.end)
        }
        GenerateType::Member => {
            let separator = if property_count > 0 {
                ","
            } else {
                ""
            };

            (separator, range.start + 1, range.start + 1)
        }
    };
```

I wouldn't attempt the original variant without Rust's "not-definitely-assigned" compiler help, but I have also made numerous mistakes with tuples recently and since Rust doesn't give me "named positions" in a tuple.

I've gone with a more-verbose tuple-assignment to protect from the user error:

```rust
    let (separator, start, end) = match generate_type {
        GenerateType::Object => {
            let separator = "";
            let start = range.start;
            let end = range.end;
            (separator, start, end)
        }
        GenerateType::Member => {
            let separator = if property_count > 0 {
                ","
            } else {
                ""
            };
            let start = range.start + 1;
            let end = range.start + 1;
            (separator, start, end)
        }
    };
```